homeassistant/components/tasmota/switch.py

```diff
@@ -0,0 +1,119 @@
+"""Support for Tasmota switches."""
+import logging
+
+from homeassistant.components import switch
+from homeassistant.components.switch import SwitchEntity
+from homeassistant.core import callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+
+from . import DOMAIN as TASMOTA_DOMAIN
+from .discovery import TASMOTA_DISCOVERY_ENTITY_NEW, clear_discovery_hash
+from .mixins import TasmotaAvailability, TasmotaDiscoveryUpdate
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up Tasmota switch dynamically through discovery."""
+
+    async def async_discover(tasmota_entity, discovery_hash):
+        """Discover and add a Tasmota switch."""
+        try:
+            async_add_entities(
+                [
+                    TasmotaSwitch(
+                        tasmota_entity=tasmota_entity, discovery_hash=discovery_hash
+                    )
+                ]
+            )
+        except Exception:
+            clear_discovery_hash(hass, discovery_hash)
+            raise
+
+    async_dispatcher_connect(
+        hass,
+        TASMOTA_DISCOVERY_ENTITY_NEW.format(switch.DOMAIN, TASMOTA_DOMAIN),
+        async_discover,
+    )
+
+
+class TasmotaSwitch(
+    TasmotaAvailability,
+    TasmotaDiscoveryUpdate,
+    SwitchEntity,
+):
+    """Representation of a Tasmota switch."""
+
+    def __init__(self, tasmota_entity, **kwds):
+        """Initialize the Tasmota switch."""
+        self._state = False
+        self._sub_state = None
+
+        self._unique_id = tasmota_entity.unique_id
+
+        super().__init__(
+            discovery_update=self.discovery_update,
+            tasmota_entity=tasmota_entity,
+            **kwds,
+        )
+
+    async def async_added_to_hass(self):
+        """Subscribe to MQTT events."""
+        await super().async_added_to_hass()
+        self._tasmota_entity.set_on_state_callback(self.state_updated)
+        await self._subscribe_topics()
+
+    async def discovery_update(self, update):
+        """Handle updated discovery message."""
+        self._tasmota_entity.config_update(update)
+        await self._subscribe_topics()
+        self.async_write_ha_state()
+
+    @callback
+    def state_updated(self, state):
+        """Handle new MQTT state messages."""
+        self._state = state
+        self.async_write_ha_state()
+
+    async def _subscribe_topics(self):
+        """(Re)Subscribe to topics."""
+        await self._tasmota_entity.subscribe_topics()
```

Comment:
Why doesn't the subscription and callback registration happen together?

Reply:
The mixins on HA side also need to register their callbacks, that's why it's kept separated.

Reply:
It's to allow the mixins to also register callbacks